R/prep_card.R

```diff
@@ -0,0 +1,330 @@
+    )
+  }
+
+  # TODO get the logic to work with strings and then add support for symbols /
+  # unquoted strings
+
+  # TODO priority for extracting context - e.g. by variables, etc:
+  #   1. direct passing of args
+  #   2. from attributes
+  #   3. tfrmt object
+
+  # column <- rlang::enquo(column)
+
+  # a <- tfrmt_find_args(..., env = environment(), parent_env = caller_env())
+  # columns_quo <- a$column |>
+  #   rlang::quos_auto_name()
+  # columns_char <- names(columns_quo)
+
+  if (has_attributes(x)) {
+    x <- x |>
+      # remove attributes for now
+      # TODO add some logic to deal with them
+      dplyr::filter(
+        .data$context != "attributes"
+      )
+  }
+
+  interim <- shuffled_card
+
+  # don't unite for hierarchical stack
+  if (!"hierarchical" %in% unique(x$context)) {
+    interim <- unite_data_vars(shuffled_card, column)
+  }
+
+  output <- interim |>
+    # process_labels() |>
+    process_big_n(column) |>
+    process_categorical_vars(column) |>
+    fill_variables(
+      variables = ard_args$variables %||% variables,
+      fill_value = fill_hierarchical_overall
+    )
+
+  output
+}
+
+# does the shuffled card have the `args` attribute - allows us to extract some
+# of the arguments of the original ard call
+has_args <- function(x) {
+  args <- attr(x, "args")
+
+  output <- !rlang::is_empty(args)
+
+  output
+}
+
+# replace_na with "Any <column-name>" where applicable (where the preceding
+# column) is not NA
+fill_variables <- function(x, variables, fill_value = "auto") {
+
+  if (length(variables) < 2) {
+    return(x)
+  }
+
+  if (fill_value == "Any {colname}") {
+    fill_value <- "auto"
+  }
+
+  pair_list <- generate_pairs(variables)
+
+  output <- x
+
+  for (i in seq_along(pair_list)) {
+    output <- replace_na_pair(output, pair_list[[i]], fill_value)
+  }
+
+  output
+}
+
+# fill_variables does pairwise conditional replacement of NAs. generate_pairs
+# builds those pairs
+generate_pairs <- function(x) {
+  # TODO drop names
+  output <- tibble::tibble(x = x) |>
+    dplyr::mutate(
+      x_lead = dplyr::lead(x)
+    ) |>
+    tidyr::drop_na() |>
+    purrr::pmap(c)
+
+  output
+}
+
+# replace missing values in one variable if a another variable is not NA
+# this is the function used by fill_variables to iterate over the pairs of
+# columns
+replace_na_pair <- function(x, pair, fill_value = "auto") {
+  variables_syms <- rlang::syms(pair)
+
+  if (fill_value == "auto") {
+    fill_value <- glue::glue("Any {variables_syms[[2]]}")
+  }
+
+  output <- x |>
+    dplyr::mutate(
+      !!variables_syms[[2]] := dplyr::if_else(
+        is.na(!!variables_syms[[2]]) & !is.na(!!variables_syms[[1]]),
+        fill_value,
+        !!variables_syms[[2]]
+      ),
+      !!variables_syms[[2]] := as.character(!!variables_syms[[2]])
+    )
+
+  output
+}
+
+# `column` here is the same value as the `column` argument
+# from `tfrmt(..., column = , ...)`
+process_big_n <- function(x, column) {
```

Comment:
if we switch `column` to `by` and this is too much, another option would be to append rows to the ARD with stat set to "bigN".

Reply:
renamed `column` to `by`.